### PR TITLE
Organize sliders into expandable summary, add first party protections section

### LINKED ISSUE #2482: Indicate when first-party protections are active
>Besides automatic tracker blocking, Privacy Badger removes outgoing link click tracking on [Facebook](https://www.eff.org/deeplinks/2018/05/privacy-badger-rolls-out-new-ways-fight-facebook-tracking) and [Google](https://www.eff.org/deeplinks/2018/10/privacy-badger-now-fights-more-sneaky-google-tracking), with more privacy protections on the way.

The thing is that we don't advertise that this happens anywhere in the UI. But we should! We should inform the user when Privacy Badger is actively removing tracking from outgoing links on Facebook and Google.

### PR DESCRIPTION
**Pull request by ablanathtanalba**
Fixes #2482.

Should help with #2021, and with the problem of users not noticing the "Disable for this site" button.

adds the following sections to the popup:
* firstparty notifications for when the current site has any unique firstparty protections protections scripts happening on it
* blocked resources container hidden by default behind the current ’n potential trackers blocked’ message

each of the sections have the option to drop down expand into a more detailed informational or control panel for each: 
* The firstparty section just drops down into a message that describes the type of protection that is taking place, and links to reading more about link tracking and how privacy badger protects against it. 
* The blocked resources hidden under the ’n potential trackers blocked’ message is the same blocked resources container the the current popup shows, with the sliders for each.

_EDIT:_ Widget section forthcoming in #2758

### CODE REVIEW

**Comment by ablanathtanalba** [comment]
Question: should I fold the work from #2744 into this PR or wait until that's merged into master, rebase, and cleanup from there?

**Comment by ghostwords** [comment]
Yes! This should replace #2744 too.

**Review by FominMax — APPROVED** [review]

**Comment by ablanathtanalba** [comment]
Screenshots of the current working state:

Screen Shot 2021-03-29 at 10 06 43 AM


Screen Shot 2021-03-29 at 10 06 55 AM

**Comment by ablanathtanalba** [comment]
Surely there's a more graceful approach to some of my css in this, but going ahead and marking ready for review now. 😅

File: src/js/popup.js
Comment (ghostwords) [inline src/js/popup.js:227]: We should only check against "matches" for "content_scripts" entries that have at least one `"/firstparties/"` script in "js".

File: src/js/popup.js
Comment (ghostwords) [inline src/js/popup.js:228]: This sort of works but could produce false positives.

What if we leave `tabHost` alone and instead `extractHostFromURL()` from "matches" entries?

- If the extracted host starts with a wildcard, see if `tabHost` ends with host (sliced to remove wildcard).
- If the host does not start with a wildcard, we should see if host and `tabHost` are exactly equal.

File: src/js/popup.js
Comment (ghostwords) [inline src/js/popup.js:217]: Let's do the work of checking if we are on a first party protections page back in the background page. The popup should just get a new boolean as part of the popup data object.

File: src/js/popup.js
Comment (ghostwords) [inline src/js/popup.js:219]: Should use `chrome.runtime.getURL()` to build extension URLs.

File: src/skin/popup.html
Comment (ghostwords) [inline src/skin/popup.html:132]: This caret is visible on pages with no third parties (`https://example.com`).

Comment (ghostwords) [reply src/skin/popup.html:132]: All carets should get the hand cursor like other clickable elements.

Comment (ghostwords) [reply src/skin/popup.html:132]: Should carets get a tooltip? "Click to show more information"?

File: src/skin/popup.html
Comment (ghostwords) [inline src/skin/popup.html:141]: I think we should center the section headers ("X potential trackers blocked", "Link tracking protection is enforced").

Comment (ghostwords) [reply src/skin/popup.html:141]: Could we make it easier to expand/collapse? The carets are pretty small click/tap targets. Maybe make it so that we could click/tap the entire header?

File: src/skin/popup.css
Comment (ghostwords) [inline src/skin/popup.css:476]: Should match separator styles (`#d3d3d3`, margins?) on the options page.

File: src/js/popup.js
Comment (ghostwords) [inline src/js/popup.js:231]: No need for this if we center the section headers.

**Review by ghostwords — COMMENTED** [review]
Thank you! Please see my first pass feedback below.

File: src/js/popup.js
Comment (ghostwords) [reply src/js/popup.js:219]: Should fetch the manifest just once and cache the result going forward.

File: src/js/webrequest.js
Comment (ghostwords) [inline src/js/webrequest.js:1140]: We could just reuse `"updateSettings"` instead.

File: src/js/webrequest.js
Comment (ghostwords) [inline src/js/webrequest.js:1041]: Link tracking protections section is in a strange state where it takes two clicks to expand it when `showExpandedTrackingSection` is `true`.

File: src/skin/popup.html
Comment (ghostwords) [reply src/skin/popup.html:132]: Whatever ends up being clickable to expand/collapse sections should be keyboard-accessible. Users should be able to <kbd>Tab</kbd> to visibly select each toggle element and <kbd>Space</kbd> (or whatever the standard key is) to activate.

File: src/js/webrequest.js
Comment (ablanathtanalba) [reply src/js/webrequest.js:1060]: I'm unable to recreate this bug :/
Are you still able to produce it?

File: src/js/webrequest.js
Comment (ablanathtanalba) [reply src/js/webrequest.js:1159]: I can easily switch to that. Is it worth breaking away from how the other similar methods in this chunk handle updating a setting though?

File: src/skin/popup.html
Comment (ablanathtanalba) [reply src/skin/popup.html]: Resolved all these except the accessible activation via the keyboard bit. At its current working state, one can focus on the element from the keyboard, but I'm unable to get the activation bit working. I tried attaching on `keydown` listeners and more tradition `onClick`s to it, no luck. On the other hand, while testing this out I realized there are a few places that aren't accessible or end up "trapping" the user into some modals that they can't easily exit out of without a mouse action. Perhaps that's worth addressing in a follow up PR?

**Review by ghostwords — COMMENTED** [review]
Thank you! Feedback for checking whether we are on a first-party protections host page, mostly.

File: src/js/webrequest.js
Comment (ghostwords) [reply src/js/webrequest.js:1159]: >Is it worth breaking away from how the other similar methods in this chunk handle updating a setting though?

Do you mean that we don't yet use the `updateSettings` message in the popup? Yeah, we don't need to add more one-off messages to toggle a setting going forward.

File: src/_locales/en_US/messages.json
Comment (ghostwords) [inline src/_locales/en_US/messages.json:649]: If we use a tooltip, it should show over the entire clickable header area, not just the carets.

File: src/skin/popup.css
Comment (ghostwords) [inline src/skin/popup.css:292]: We don't want a hand cursor in the expanded section, just over the clickable header.

File: src/skin/popup.css
Comment (ghostwords) [inline src/skin/popup.css:295]: "No trackers blocked" on a page with no third-party resources (https://example.com) and "Privacy Badger is disabled on this site" also incorrectly get the hand cursor.

File: src/_locales/en_US/messages.json
Comment (ghostwords) [inline src/_locales/en_US/messages.json:425]: By the way, this link doesn't work because it doesn't have `target="_blank"`.

File: src/skin/popup.html
Comment (ghostwords) [inline src/skin/popup.html:134]: Collapsing shouldn't get the same tooltip.

Maybe instead of tooltips, we should add a hover effect to make it more clear you can click on the header to get more information. Maybe the caret should move a little.

File: src/skin/popup.html
Comment (ghostwords) [reply src/skin/popup.html]: `<div>` or `<span>` elements don't have built-in keyboard accessibility; '<a>' and `<button>` elements do.

To make a `<span>` focusable and so able to receive keyboard events, you have to add a `tabindex=0`. You then have to add a `keydown` listener and filter for supported keys.

Make sure to add the `tabindex` property to the same element that your `keydown` listener is on.

Comment (ghostwords) [reply src/skin/popup.html]: You'll also want to add proper ARIA attributes (such as `role=button` perhaps) to your elements.

Comment (ghostwords) [reply src/skin/popup.html]: You could instead use a `<button>` or an `<a href="" role="button">` element, which automatically support keyboard navigation and also trigger your click handler on <kbd>Space</kbd>/<kbd>Enter</kbd> presses. You will probably want to `e.preventDefault()` inside the click handler (to avoid performing the unwanted default action of `<a>`).

Comment (ghostwords) [reply src/skin/popup.html]: Yes to improving keyboard navigation in a followup PR: https://github.com/EFForg/privacybadger/pull/2629#issuecomment-646101693

File: src/skin/popup.html
Comment (ghostwords) [reply src/skin/popup.html:134]: Yeah, let's do a hover effect over the clickable area. More clear/helpful than tooltips, no translation required.

File: src/js/background.js
Comment (ablanathtanalba) [reply src/js/background.js:111]: Since the `firstPartyProtectionsEnabled()` method got moved to utils and now only the `badger.firstPartiesList` gets set on startup, I think that resolves this comment, yes?

Comment (ghostwords) [reply src/js/background.js:111]: Well, do we need to collect match patterns of first party protection scripts on Badger startup? If we don't, let's have `badger.firstPartiesList` initialization happen the first time we need it. My code sample above illustrates an approach that accomplishes this.

File: src/js/webrequest.js
Comment (ablanathtanalba) [reply src/js/webrequest.js]: Ah gotcha, fixed this

File: src/js/background.js
Comment (ablanathtanalba) [reply src/js/background.js]: I didn't follow the exact pattern above, but stuck to the same idea on fetching the firstparties list only the first time its needed rather than during startup. Thank you!

File: src/js/utils.js
Comment (ghostwords) [inline src/js/utils.js:479]: Why assume tab host starts with "www"? What about, for example, `m.facebook.com`?

File: src/js/background.js
Comment (ghostwords) [reply src/js/background.js]: We don't need `badger.firstPartiesList` though anywhere else though, and we shouldn't have to pass in the `badger` object with every call to this function either. Keeping local state is what the `firstPartiesList` local-to-the-function-[closure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) variable in the code snippet above accomplishes.

File: src/js/utils.js
Comment (ghostwords) [inline src/js/utils.js:476]: Nit, but: "avoid false positives" is something you'd write in the body of your commit message that improves some check to avoid false positives. It doesn't make sense as a comment though, I don't think, as it refers to something that's irrelevant at this point.

File: src/js/background.js
Comment (ablanathtanalba) [reply src/js/background.js]: Oh, right! Good call, then. I've changed the method to that, then.

**Comment by ablanathtanalba** [comment]
The tweaks you left @ghostwords look good! I like the styling you added too, with the orange outline and expanding animation

File: src/js/utils.js
Comment (ghostwords) [inline src/js/utils.js:482]: You shouldn't have to assume anything about what tab host subdomains look like. Does the tab host end with the wildcard pattern minus the wildcard?

**Pull request merged**
